python/services/dataservice/dmod/dataservice/service.py

```diff
@@ -716,23 +868,107 @@ async def can_be_fulfilled(self, requirement: DataRequirement) -> Tuple[bool, Op
+        ----------
+        requirement : DataRequirement
+            The requirement for which the capability to derive a realization config needs to be determined.
+        job : Job
+            The job having the given requirement.
+
+        Returns
+        -------
+        bool
+            Whether deriving an appropriate realization configuration is possible.
+        """
+        if requirement.category != DataCategory.CONFIG:
+            return False
+        elif requirement.domain.data_format != DataFormat.NGEN_REALIZATION_CONFIG:
+            return False
+
+        request = job.model_request
+        return (isinstance(request, AbstractNgenRequest) and request.is_intelligent_request
+                and isinstance(request.formulation_configs, PartialRealizationConfig))
+
+    async def derive_datasets(self, job: Job) -> List[DataRequirement]:
+        """
+        Derive any datasets as required for the given job awaiting its data.
+
+        Job is expected to be in the ``AWAITING_DATA`` status step.  If it is not, no datasets are derived, a warning is
+        logged, and an empty list is returned.
+
+        If in the right status, but initially any unfulfilled requirements of the job cannot have a dataset successfully
+        derived, a ::class:`DmodRuntimeError` is raised.
+
+        Parameters
+        ----------
+        job
+
+        Returns
+        -------
+        List[DataRequirement]
+            A list of the given job's data requirements for a which a fulfilling dataset was derived and associated.
+
+        Raises
+        -------
+        DmodRuntimeError
+            Raised if a job with the correct status has an initially unfulfilled requirement for which a satisfactory
+            dataset can not be derived by this function.
+        """
+        # Only do something if the job has the right status
+        if job.status_step != JobExecStep.AWAITING_DATA:
+            return []
+
+        results = []
+
+        for req in [r for r in job.data_requirements if r.fulfilled_by is None]:
+            # **********************************************************************************************************
+            # *** NOTE: if/when deriving forcing datasets is supported, make sure this is done before config datasets
+            # **********************************************************************************************************
+            # Derive realization config datasets from formulations in message body when necessary
+            if req.category == DataCategory.CONFIG and req.domain.data_format == DataFormat.NGEN_REALIZATION_CONFIG:
+                self._derive_realization_config_from_formulations(requirement=req, job=job)
+                results.append(req)
+            # The above are the only supported derivations, so blow up here if there was something else
+            else:
+                msg_template = "Unsupported requirement dataset derivation for job {} (requirement: {})"
```

Comment:
It seems like we either want to batch create or roll back here. I am assuming that it is possible to successfully create a dataset in a prior iteration and then fail in the next. Maybe its just a difference of expectations, but I would expect that the prior created datasets would be removed. Thoughts?

Reply:
The idea I had in mind was for some datasets to be temporary with an expiration time.  The data service would periodically clean up these.

I expected that some derived datasets may not be trivial to reproduce - derived forcings, for example - so we might want to hang onto them a little while after initially appearing to be done with them.  It also seemed likely that any datasets that were trivial to reproduce would also be trivial to store for a bit.

Of course, things are not fully implemented to make that work, but they've been started, and I added some more TODO comments around this area to help with that.

Reply:
That makes a lot of sense. I wasnt thinking about this in the context of long running tasks, like deriving forcing. Thanks for sharing your thoughts.